### ISSUE TITLE
Idea: string_as_ptr_for_ffi

### ISSUE THREAD

**Issue by sevenc-nanashi** [rust-lang/rust-clippy#16021]
### What it does

This lint will warn passing `*const u8` or `*const i8` that is created by `String::as_ptr`, to `extern "C"` functions.
```rs
extern "C" {
    fn some_c_lib_that_takes_a(string: *const u8);
}

let string = String::from("hello world!");
some_c_lib_that_takes_a(string.as_ptr());
```

This does not work as you expected, as `String::as_ptr` does not return C-style (null-terminated) string pointer.

### Advantage

- This prevents forgetting to wrap the string using CString.

### Drawbacks

- Some libraries actually wants this behavior, so it might cause warning on existing code.

### Example

```rust
extern "C" {
    fn some_c_lib_that_takes_a(string: *const u8);
}

let string = String::from("hello world!");
some_c_lib_that_takes_a(string.as_ptr());
```

Could be written as:

```rust
extern "C" {
    fn some_c_lib_that_takes_a(string: *const u8);
}

let string = String::from("hello world!");
let c_string = std::ffi::CString.new(&string).unwrap();
some_c_lib_that_takes_a(c_string.as_ptr());
```

### Comparison with existing lints

I couldn't find existing.

### Additional Context

N/A
